src/Application/Filters/Actions/ActionWatershedFilter.cc

```diff
@@ -0,0 +1,427 @@
+	//std::vector<LayerHandle> dst_layer_;
+	LayerHandle dst_layer_;
+
+	double watershedThreshold_val_;
+	double watershedLevel_val_;
+
+	std::vector<std::string> dst_layer_ids_;
+	std::vector<_LabelType> watershed_image_output_labels_;
+  
+public:
+	// RUN:
+	// Implemtation of run of the Runnable base class, this function is called when the thread
+	// is launched.
+
+	// NOTE: The macro needs a data type to select which version to run. This needs to be
+	// a member variable of the algorithm class.
+	SCI_BEGIN_TYPED_ITK_RUN( this->src_layer_->get_data_type() )
+	{
+		// Define the type of filter that we use.
+		typedef itk::WatershedImageFilter< TYPED_IMAGE_TYPE > filter_type;
+		
+		// Retrieve the image as an itk image from the underlying data structure
+		// NOTE: This only does wrapping and does not regenerate the data.
+		typename Core::ITKImageDataT<VALUE_TYPE>::Handle input_image; 
+		this->get_itk_image_from_layer<VALUE_TYPE>( this->src_layer_, input_image );
+		
+		// Create a new ITK filter instantiation.	
+		typename filter_type::Pointer filter = filter_type::New();
+		
+		// Relay abort and progress information to the layer that is executing the filter.
+		this->forward_abort_to_filter( filter, this->dst_layer_ );
+		this->observe_itk_progress( filter, this->dst_layer_ );
+		
+		filter->SetLevel( this->watershedLevel_val_ );
+		filter->SetThreshold( this->watershedThreshold_val_ );
+		filter->SetInput( input_image->get_image() );
+		
+		// Ensure we will have some threads left for doing something else
+		this->limit_number_of_itk_threads( filter );
+		
+		// Run the actual ITK filter.
+		// This needs to be in a try/catch statement as certain filters throw exceptions when they
+		// are aborted. In that case we will relay a message to the status bar for information.
+#if (_ConsoleOuput == true)
+		std::cout << "\nRunning watershed filter:\n";
+		std::cout << "\tPROGRESS :: Updating watershedding filter...\n";
+#endif
+		try 
+		{ 
+			filter->Update(); 
+		} 
+		catch ( ... ) 
+		{
+			if ( this->check_abort() )
+			{
+				this->report_error( "Filter was aborted." );
+				return;
+			}
+			this->report_error( "ITK filter failed to complete." );
+			return;		
+		}
+		
+		// As ITK filters generate an inconsistent abort behavior, we record our own abort flag
+		// This one is set when the abort button is pressed and an abort is sent to ITK.
+		if ( this->check_abort() ) return;
+#if (_ConsoleOuput == true)
+		std::cout << "\tPROGRESS :: Casting filter to output type...\n";
+#endif
+		typedef itk::CastImageFilter< _ImageTypeFilter, _ImageTypeOut > CastImageFilter_FilterToOut;
+		CastImageFilter_FilterToOut::Pointer castImage_FilterToOut = CastImageFilter_FilterToOut::New();
+		castImage_FilterToOut->SetInput( filter->GetOutput() );
+		castImage_FilterToOut->Update();
+#if (_ConsoleOuput == true)
+		std::cout << "\tPROGRESS :: Reading in all labels (extreme patience required)...\n";
+#endif
+#if (_Timing == true)
```

Comment:
Do you mean use a preprocessor define of _Debug to encompass both _Timing and _ConsoleOutput?  Or, better yet, how do I check if the system was built as a debug version?